crates/sui-move-build/src/linters/share_owned.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This analysis flags making objects passed as function parameters or resulting from unpacking
+//! (likely already owned) shareable which would lead to an abort. A typical patterns is to create a
+//! fresh object and share it within the same function
+
+use move_ir_types::location::*;
+
+use move_compiler::{
+    cfgir::{
+        absint::JoinResult,
+        ast::Program,
+        visitor::{
+            LocalState, SimpleAbsInt, SimpleAbsIntConstructor, SimpleDomain, SimpleExecutionContext,
+        },
+        CFGContext,
+    },
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        Diagnostic, Diagnostics,
+    },
+    hlir::ast::{
+        BaseType_, Command, Exp, LValue, LValue_, Label, ModuleCall, SingleType, SingleType_, Type,
+        Type_, Var,
+    },
+    parser::ast::Ability_,
+    shared::Identifier,
+};
+use std::collections::BTreeMap;
+
+const SHARE_FUNCTIONS: &[(&str, &str, &str)] = &[
+    ("sui", "transfer", "public_share_object"),
+    ("sui", "transfer", "share_object"),
+];
+
+const SHARE_OWNED_DIAG: DiagnosticInfo = custom(
+    "Lint ",
+    Severity::Warning,
+    /* category */ 1,
+    /* code */ 1,
+    "possible owned object share",
+);
```

Comment:
Any feedback you have for making this easier to manage? (specifically on the category/code side of things)

Reply:
Not at the moment, but I will think about it when writing more lints!